File: backend/src/services/queue/EventQueueService.ts
import { v4 as uuidv4 } from 'uuid';
import { db } from '../../db/database';
import { logger } from '../../utils/logger';
import { broadcast } from '../../utils/websocket';
import { config } from '../../config';

export interface QueuedEvent {
  id: string;
  type: 'MACHINE_STATUS' | 'COUNTER';
  equipment_id: string;
  equipment_name: string;
  payload: Record<string, unknown>;
  status: 'PENDING' | 'SENDING' | 'SENT' | 'FAILED';
  retry_count: number;
  next_retry_at?: number;
  created_at: number;
  sent_at?: number;
  error_message?: string;
}

export interface QueueStats {
  pending: number;
  sending: number;
  sent: number;
  failed: number;
  total: number;
}

// Exponential backoff delays in milliseconds (indexed by retry_count)
const RETRY_DELAYS_MS = [5000, 15000, 30000, 60000, 120000, 300000, 300000, 300000, 300000, 300000];

export class EventQueueService {
  private processingLoopTimer: NodeJS.Timeout | null = null;
  private cleanupTimer: NodeJS.Timeout | null = null;
  private aprisoService: { publishMachineStatus(event: QueuedEvent): Promise<void>; publishCounter(event: QueuedEvent): Promise<void> } | null = null;

  setAprisoService(svc: { publishMachineStatus(event: QueuedEvent): Promise<void>; publishCounter(event: QueuedEvent): Promise<void> }): void {
    this.aprisoService = svc;
  }

  enqueue(event: Omit<QueuedEvent, 'id' | 'status' | 'retry_count' | 'created_at'>): void {
    const id = uuidv4();
    const now = Math.floor(Date.now() / 1000);

    const stmt = db.prepare(`
      INSERT INTO event_queue (id, type, equipment_id, equipment_name, payload, status, retry_count, created_at)
      VALUES (?, ?, ?, ?, ?, 'PENDING', 0, ?)
    `);

    stmt.run(id, event.type, event.equipment_id, event.equipment_name, JSON.stringify(event.payload), now);

    logger.info(`Event enqueued: ${id} type=${event.type} equipment=${event.equipment_name}`);

    broadcast('event_queued', {
      id,
      type: event.type,
      equipmentId: event.equipment_id,
      equipmentName: event.equipment_name,
      status: 'PENDING',
    });
  }

  async processQueue(): Promise<void> {
    if (!this.aprisoService) {
      logger.warn('EventQueueService: AprisoService not set, skipping queue processing');
      return;
    }

    const now = Math.floor(Date.now() / 1000);

    // Fetch PENDING events that are ready to process
    const events = db.prepare(`
      SELECT * FROM event_queue
      WHERE status = 'PENDING'
        AND (next_retry_at IS NULL OR next_retry_at <= ?)
      ORDER BY created_at ASC
      LIMIT 50
    `).all(now) as QueuedEvent[];

    if (events.length === 0) return;

    logger.debug(`Processing ${events.length} queued events`);

    for (const event of events) {
      await this.processEvent(event);
    }
  }

  private async processEvent(event: QueuedEvent): Promise<void> {
    if (!this.aprisoService) return;

    // Mark as SENDING
    db.prepare(`UPDATE event_queue SET status = 'SENDING' WHERE id = ?`).run(event.id);
    broadcast('event_status_changed', { id: event.id, status: 'SENDING' });

    try {
      if (event.type === 'MACHINE_STATUS') {
        await this.aprisoService.publishMachineStatus(event);
      } else if (event.type === 'COUNTER') {
        await this.aprisoService.publishCounter(event);
      }

      const sentAt = Math.floor(Date.now() / 1000);
      db.prepare(`
        UPDATE event_queue SET status = 'SENT', sent_at = ?, error_message = NULL WHERE id = ?
      `).run(sentAt, event.id);

      logger.info(`Event sent successfully: ${event.id} type=${event.type} equipment=${event.equipment_name}`);
      broadcast('event_status_changed', { id: event.id, status: 'SENT', sentAt });

    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      const newRetryCount = event.retry_count + 1;

      if (newRetryCount >= config.queue.maxRetries) {
        db.prepare(`
          UPDATE event_queue
          SET status = 'FAILED', retry_count = ?, error_message = ?, next_retry_at = NULL
          WHERE id = ?
        `).run(newRetryCount, errorMessage, event.id);

        logger.error(`Event permanently failed after ${newRetryCount} retries: ${event.id} - ${errorMessage}`);
        broadcast('event_status_changed', { id: event.id, status: 'FAILED', retryCount: newRetryCount, error: errorMessage });
      } else {
        const delayMs = RETRY_DELAYS_MS[Math.min(event.retry_count, RETRY_DELAYS_MS.length - 1)];
        const nextRetryAt = Math.floor((Date.now() + delayMs) / 1000);

        db.prepare(`
          UPDATE event_queue
          SET status = 'PENDING', retry_count = ?, next_retry_at = ?, error_message = ?
          WHERE id = ?
        `).run(newRetryCount, nextRetryAt, errorMessage, event.id);

        logger.warn(`Event retry scheduled: ${event.id} attempt=${newRetryCount} next_retry_in=${delayMs}ms - ${errorMessage}`);
        broadcast('event_status_changed', {
          id: event.id,
          status: 'PENDING',
          retryCount: newRetryCount,
          nextRetryAt,
          error: errorMessage,
        });
      }
    }
  }

  startProcessingLoop(): void {
    const intervalMs = config.queue.processingIntervalMs;
    logger.info(`Starting event queue processing loop (interval: ${intervalMs}ms)`);

    this.processingLoopTimer = setInterval(() => {
      this.processQueue().catch((err) => {
        logger.error(`Queue processing loop error: ${err instanceof Error ? err.message : String(err)}`);
      });
    }, intervalMs);

    // Run cleanup on startup
    this.cleanup();

    // Schedule cleanup every hour
    this.cleanupTimer = setInterval(() => {
      this.cleanup();
    }, 60 * 60 * 1000);
  }

  stopProcessingLoop(): void {
    if (this.processingLoopTimer) {
      clearInterval(this.processingLoopTimer);
      this.processingLoopTimer = null;
    }
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
  }

  cleanup(): void {
    const retentionSeconds = config.queue.retentionHours * 3600;
    const cutoffTime = Math.floor(Date.now() / 1000) - retentionSeconds;

    const result = db.prepare(`
      DELETE FROM event_queue WHERE status = 'SENT' AND sent_at < ?
    `).run(cutoffTime);

    if (result.changes > 0) {
      logger.info(`Queue cleanup: removed ${result.changes} old SENT events`);
    }
  }

  getStats(): QueueStats {
    const rows = db.prepare(`
      SELECT status, COUNT(*) as count FROM event_queue GROUP BY status
    `).all() as Array<{ status: string; count: number }>;

    const stats: QueueStats = { pending: 0, sending: 0, sent: 0, failed: 0, total: 0 };

    for (const row of rows) {
      const status = row.status.toLowerCase() as keyof Omit<QueueStats, 'total'>;
      if (status in stats) {
        stats[status] = row.count;
      }
      stats.total += row.count;
    }

    return stats;
  }
}

export const eventQueueService = new EventQueueService();

File: backend/services/queue/event_queue_service.py
import asyncio
import json
import time
import uuid
from typing import Any

from config import config
from db import db_lock, get_db
from utils.logger import logger
from utils.websocket import broadcast

# Exponential backoff delays in seconds (indexed by retry_count)
RETRY_DELAYS = [5, 15, 30, 60, 120, 300, 300, 300, 300, 300]


class EventQueueService:
    def __init__(self) -> None:
        self._apriso_service: Any = None
        self._processing_task: asyncio.Task | None = None
        self._cleanup_task: asyncio.Task | None = None

    def set_apriso_service(self, svc: Any) -> None:
        self._apriso_service = svc

    def enqueue(self, event_type: str, equipment_id: str, equipment_name: str, payload: dict) -> None:
        event_id = str(uuid.uuid4())
        now = int(time.time())
        db = get_db()
        with db_lock():
            db.execute(
                """
                INSERT INTO event_queue
                    (id, type, equipment_id, equipment_name, payload, status, retry_count, created_at)
                VALUES (?, ?, ?, ?, ?, 'PENDING', 0, ?)
                """,
                (event_id, event_type, equipment_id, equipment_name, json.dumps(payload), now),
            )
            db.commit()
        logger.info(f"Event enqueued: {event_id} type={event_type} equipment={equipment_name}")
        asyncio.get_event_loop().create_task(
            broadcast("event_queued", {
                "id": event_id,
                "type": event_type,
                "equipmentId": equipment_id,
                "equipmentName": equipment_name,
                "status": "PENDING",
            })
        )

    async def process_queue(self) -> None:
        if self._apriso_service is None:
            logger.warning("EventQueueService: AprisoService not set, skipping")
            return

        now = int(time.time())
        db = get_db()
        rows = db.execute(
            """
            SELECT * FROM event_queue
            WHERE status = 'PENDING'
              AND (next_retry_at IS NULL OR next_retry_at <= ?)
            ORDER BY created_at ASC
            LIMIT 50
            """,
            (now,),
        ).fetchall()

        if not rows:
            return

        logger.debug(f"Processing {len(rows)} queued events")
        for row in rows:
            await self._process_event(dict(row))

    async def _process_event(self, event: dict) -> None:
        db = get_db()
        with db_lock():
            db.execute("UPDATE event_queue SET status = 'SENDING' WHERE id = ?", (event["id"],))
            db.commit()
        await broadcast("event_status_changed", {"id": event["id"], "status": "SENDING"})

        try:
            payload = json.loads(event["payload"])
            event_with_payload = {**event, "payload": payload}

            if event["type"] == "MACHINE_STATUS":
                await self._apriso_service.publish_machine_status(event_with_payload)
            elif event["type"] == "COUNTER":
                await self._apriso_service.publish_counter(event_with_payload)

            sent_at = int(time.time())
            with db_lock():
                db.execute(
                    "UPDATE event_queue SET status = 'SENT', sent_at = ?, error_message = NULL WHERE id = ?",
                    (sent_at, event["id"]),
                )
                db.commit()
            logger.info(f"Event sent: {event['id']} type={event['type']} equipment={event['equipment_name']}")
            await broadcast("event_status_changed", {"id": event["id"], "status": "SENT", "sentAt": sent_at})

        except Exception as exc:
            error_message = str(exc)
            new_retry_count = event["retry_count"] + 1

            if new_retry_count >= config.queue.max_retries:
                with db_lock():
                    db.execute(
                        """UPDATE event_queue
                           SET status = 'FAILED', retry_count = ?, error_message = ?, next_retry_at = NULL
                           WHERE id = ?""",
                        (new_retry_count, error_message, event["id"]),
                    )
                    db.commit()
                logger.error(
                    f"Event permanently failed after {new_retry_count} retries: {event['id']} - {error_message}"
                )
                await broadcast("event_status_changed", {
                    "id": event["id"], "status": "FAILED",
                    "retryCount": new_retry_count, "error": error_message,
                })
            else:
                delay = RETRY_DELAYS[min(event["retry_count"], len(RETRY_DELAYS) - 1)]
                next_retry_at = int(time.time()) + delay
                with db_lock():
                    db.execute(
                        """UPDATE event_queue
                           SET status = 'PENDING', retry_count = ?, next_retry_at = ?, error_message = ?
                           WHERE id = ?""",
                        (new_retry_count, next_retry_at, error_message, event["id"]),
                    )
                    db.commit()
                logger.warning(
                    f"Event retry scheduled: {event['id']} attempt={new_retry_count} "
                    f"next_retry_in={delay}s - {error_message}"
                )
                await broadcast("event_status_changed", {
                    "id": event["id"], "status": "PENDING",
                    "retryCount": new_retry_count,
                    "nextRetryAt": next_retry_at,
                    "error": error_message,
                })

    def start_processing_loop(self) -> None:
        interval_s = config.queue.processing_interval_ms / 1000
        logger.info(f"Starting event queue processing loop (interval: {interval_s}s)")

        async def _loop() -> None:
            while True:
                await asyncio.sleep(interval_s)
                try:
                    await self.process_queue()
                except Exception as exc:
                    logger.error(f"Queue processing loop error: {exc}")

        async def _cleanup_loop() -> None:
            self.cleanup()
            while True:
                await asyncio.sleep(3600)
                self.cleanup()

        self._processing_task = asyncio.get_event_loop().create_task(_loop())
        self._cleanup_task = asyncio.get_event_loop().create_task(_cleanup_loop())

    def stop_processing_loop(self) -> None:
        if self._processing_task:
            self._processing_task.cancel()
            self._processing_task = None
        if self._cleanup_task:
            self._cleanup_task.cancel()
            self._cleanup_task = None

    def cleanup(self) -> None:
        cutoff = int(time.time()) - config.queue.retention_hours * 3600
        db = get_db()
        with db_lock():
            cursor = db.execute(
                "DELETE FROM event_queue WHERE status = 'SENT' AND sent_at < ?", (cutoff,)
            )
            db.commit()
        if cursor.rowcount > 0:
            logger.info(f"Queue cleanup: removed {cursor.rowcount} old SENT events")

    def get_stats(self) -> dict:
        db = get_db()
        rows = db.execute("SELECT status, COUNT(*) as count FROM event_queue GROUP BY status").fetchall()
        stats = {"pending": 0, "sending": 0, "sent": 0, "failed": 0, "total": 0}
        for row in rows:
            key = row["status"].lower()
            if key in stats:
                stats[key] = row["count"]
            stats["total"] += row["count"]
        return stats


event_queue_service = EventQueueService()
